src/libs/sys-util-rs/src/cgroup.rs

```diff
@@ -0,0 +1,718 @@
+    } else {
+        Ok(cgroups::Cgroup::new(hie, valid_path.as_str()))
+    }
+}
+
+/// Check whether `path` hosts a cgroup hierarchy directory.
+pub fn is_cgroup_exist(path: &str) -> bool {
+    let valid_path = match valid_cgroup_path(path) {
+        Ok(v) => v,
+        Err(e) => {
+            warn!(sl!(), "{}", e);
+            return false;
+        }
+    };
+
+    let v1 = get_cgroup_hierarchies().clone();
+    let cg = cgroups::Cgroup::load(Box::new(v1), valid_path.as_str());
+    for subsys in cg.subsystems() {
+        if subsys.to_controller().exists() {
+            debug!(sl!(), "cgroup {} exist", path);
+            return true;
+        }
+    }
+
+    false
+}
+
+// Validate the cgroup path is a relative path, do not include ".", "..".
+fn valid_cgroup_path(path: &str) -> Result<String> {
+    let path = path.trim_start_matches('/').to_string();
+
+    for comp in Path::new(&path).components() {
+        if !matches!(comp, Component::Normal(_)) {
+            return Err(Error::InvalidCgroupPath(path.to_string()));
+        }
+    }
+
+    Ok(path)
+}
+
+/// Remove all task from cgroup and delete the cgroup.
+pub fn force_delete_cgroup(cg: cgroups::Cgroup) -> Result<()> {
+    delete_cgroup_with_retry(cg, |cg: &Cgroup| {
+        // if task exist need to delete first.
+        for cg_pid in cg.tasks() {
+            warn!(sl!(), "Delete cgroup task pid {}", cg_pid.pid);
+            cg.remove_task(cg_pid);
+        }
+    })
+}
+
+/// Try to delete a cgroup, call the `do_process` handler at each iteration.
+pub fn delete_cgroup_with_retry<F>(cg: Cgroup, mut do_process: F) -> Result<()>
+where
+    F: FnMut(&Cgroup),
+{
+    // sleep DURATION
+    const DURATION: u64 = 10;
+    const RETRY: u64 = 200;
+
+    // In case of deletion failure caused by "Resource busy", sleep DURATION and retry RETRY times.
+    for index in 0..RETRY {
+        do_process(&cg);
+
+        if cg.delete().is_ok() {
+            if index > 0 {
+                info!(
+                    sl!(),
+                    "cgroup delete cgroup cost {} ms, retry {} times",
+                    index * DURATION,
+                    index,
+                );
+            }
+            return Ok(());
+        }
+        std::thread::sleep(std::time::Duration::from_millis(DURATION))
+    }
+
+    Err(Error::DeleteCgroup(RETRY))
+}
+
+/// Move the process `pid` into the cgroup `to`.
+pub fn move_tgid(pid: u64, to: &Cgroup) -> Result<()> {
+    info!(sl!(), "try to move tid {:?}", pid);
+    to.add_task_by_tgid(CgroupPid::from(pid))
+        .map_err(|e| Error::AddTgid(pid, e))
+}
+
+/// Move all processes tasks from `from` to `to`.
+pub fn move_cgroup_task(from: &Cgroup, to: &Cgroup) -> Result<()> {
+    info!(sl!(), "try to move tasks {:?}", from.tasks());
+    for cg_pid in from.tasks() {
+        from.remove_task(CgroupPid::from(cg_pid.pid));
+        // TODO: enhance cgroups to implement Copy for CgroupPid
```

Comment:
If would be great if you could raise GH issues and reference them for all the TODO and FIXME's.

Reply:
I have opened an issued for it, but missed the reference here.